### Assignment 4

I am setting alpha = 0.05

In [71]:
%matplotlib inline
import pylab as pl
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
%matplotlib inline

if os.getenv ('PUI2018') is None:
    print ("Must set env variable PUI2018")
if os.getenv ('PUIDATA') is None:
    print ("Must set env variable PUI2018")

import os
import json


Must set env variable PUI2018


In [72]:
jan_df = pd.read_csv('%s/JC-201801-citibike-tripdata.csv.zip'%os.getenv('PUIDATA'))
aug_df = pd.read_csv('%s/JC-201808-citibike-tripdata.csv.zip'%os.getenv('PUIDATA'))

In [73]:
df = pd.concat([jan_df, aug_df], axis=0)
df.head()

tripduration                 starttime                  stoptime  \
0           932  2018-01-01 02:06:17.5410  2018-01-01 02:21:50.0270   
1           550  2018-01-01 12:06:18.0390  2018-01-01 12:15:28.4430   
2           510  2018-01-01 12:06:56.9780  2018-01-01 12:15:27.8100   
3           354  2018-01-01 14:53:10.1860  2018-01-01 14:59:05.0960   
4           250  2018-01-01 17:34:30.1920  2018-01-01 17:38:40.9840   

   start station id start station name  start station latitude  \
0              3183     Exchange Place               40.716247   
1              3183     Exchange Place               40.716247   
2              3183     Exchange Place               40.716247   
3              3183     Exchange Place               40.716247   
4              3183     Exchange Place               40.716247   

   start station longitude  end station id end station name  \
0               -74.033459            3199     Newport Pkwy   
1               -74.033459            3199     Newport Pkwy   
2               -74.033459            3199     Newport Pkwy   
3               -74.033459            3267     Morris Canal   
4               -74.033459            3639       Harborside   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.728745             -74.032108   31929  Subscriber   
1             40.728745             -74.032108   31845  Subscriber   
2             40.728745             -74.032108   31708  Subscriber   
3             40.712419             -74.038526   31697  Subscriber   
4             40.719252             -74.034234   31861  Subscriber   

   birth year  gender  
0        1992       1  
1        1969       2  
2        1946       1  
3        1994       1  
4        1991       1

In [74]:
df.describe()

tripduration  start station id  start station latitude  \
count  5.710900e+04      57109.000000            57109.000000   
mean   6.614016e+02       3263.899473               40.722645   
std    1.023374e+04        137.335992                0.007169   
min    6.100000e+01       3183.000000               40.692640   
25%    2.300000e+02       3187.000000               40.717733   
50%    3.340000e+02       3205.000000               40.721525   
75%    5.410000e+02       3272.000000               40.727224   
max    2.061932e+06       3694.000000               40.748716   

       start station longitude  end station id  end station latitude  \
count             57109.000000    57109.000000          57109.000000   
mean                -74.045769     3257.742387             40.722275   
std                   0.010710      144.334591              0.007023   
min                 -74.096937      127.000000             40.685396   
25%                 -74.050389     3186.000000             40.717733   
50%                 -74.043117     3203.000000             40.721124   
75%                 -74.038051     3270.000000             40.727224   
max                 -74.032108     3694.000000             40.806758   

       end station longitude        bikeid    birth year        gender  
count           57109.000000  57109.000000  57109.000000  57109.000000  
mean              -74.045348  29723.676916   1980.501952      1.156438  
std                 0.010767   2512.626649     10.311655      0.501877  
min               -74.096937  15114.000000   1887.000000      0.000000  
25%               -74.049968  29207.000000   1974.000000      1.000000  
50%               -74.043117  29550.000000   1983.000000      1.000000  
75%               -74.037683  31760.000000   1988.000000      1.000000  
max               -73.960708  33695.000000   2002.000000      2.000000

In [75]:
df = df.dropna()

In [76]:
# Clean data so trips are no longer than an hour
df = df[df['tripduration'] < 60*60]

In [77]:
df.describe()

tripduration  start station id  start station latitude  \
count  56574.000000      56574.000000            56574.000000   
mean     472.606126       3263.944922               40.722669   
std      439.702037        137.362426                0.007156   
min       61.000000       3183.000000               40.692640   
25%      229.000000       3187.000000               40.717733   
50%      332.000000       3205.000000               40.721525   
75%      532.000000       3272.000000               40.727224   
max     3599.000000       3694.000000               40.748716   

       start station longitude  end station id  end station latitude  \
count             56574.000000    56574.000000          56574.000000   
mean                -74.045761     3257.660887             40.722289   
std                   0.010699      144.316993              0.007002   
min                 -74.096937      127.000000             40.685396   
25%                 -74.050389     3186.000000             40.717733   
50%                 -74.043117     3203.000000             40.721124   
75%                 -74.038051     3270.000000             40.727224   
max                 -74.032108     3694.000000             40.774667   

       end station longitude        bikeid    birth year        gender  
count           56574.000000  56574.000000  56574.000000  56574.000000  
mean              -74.045334  29727.245130   1980.537367      1.160056  
std                 0.010748   2510.378823     10.304498      0.497172  
min               -74.096937  15114.000000   1887.000000      0.000000  
25%               -74.049968  29208.000000   1974.000000      1.000000  
50%               -74.043117  29550.000000   1983.000000      1.000000  
75%               -74.037683  31760.000000   1988.000000      1.000000  
max               -73.974315  33695.000000   2001.000000      2.000000

## Comparing Day and Night trip durations

In [78]:
df.head()

tripduration                 starttime                  stoptime  \
0           932  2018-01-01 02:06:17.5410  2018-01-01 02:21:50.0270   
1           550  2018-01-01 12:06:18.0390  2018-01-01 12:15:28.4430   
2           510  2018-01-01 12:06:56.9780  2018-01-01 12:15:27.8100   
3           354  2018-01-01 14:53:10.1860  2018-01-01 14:59:05.0960   
4           250  2018-01-01 17:34:30.1920  2018-01-01 17:38:40.9840   

   start station id start station name  start station latitude  \
0              3183     Exchange Place               40.716247   
1              3183     Exchange Place               40.716247   
2              3183     Exchange Place               40.716247   
3              3183     Exchange Place               40.716247   
4              3183     Exchange Place               40.716247   

   start station longitude  end station id end station name  \
0               -74.033459            3199     Newport Pkwy   
1               -74.033459            3199     Newport Pkwy   
2               -74.033459            3199     Newport Pkwy   
3               -74.033459            3267     Morris Canal   
4               -74.033459            3639       Harborside   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.728745             -74.032108   31929  Subscriber   
1             40.728745             -74.032108   31845  Subscriber   
2             40.728745             -74.032108   31708  Subscriber   
3             40.712419             -74.038526   31697  Subscriber   
4             40.719252             -74.034234   31861  Subscriber   

   birth year  gender  
0        1992       1  
1        1969       2  
2        1946       1  
3        1994       1  
4        1991       1

In [79]:
# I am defining a 'Day' trip as any trip that starts during the time between 6 a.m. to 6 p.m. and a 'Night' trip 
# as any trip that starts during the time between 6 p.m. and 6 a.m.

# Create variable if trip is day or night 
df['dayNight'] = df['starttime'].apply(lambda x: 'Day' if ((int(x.split(' ')[1].split(':')[0]) < 18) & (int(x.split(' ')[1].split(':')[0]) >= 6)) else 'Night')

In [80]:
df['dayTripDuration'] = np.where(df['dayNight'] == 'Day', df['tripduration'], np.NaN)
df['nightTripDuration'] = np.where(df['dayNight'] == 'Night', df['tripduration'], np.NaN)

In [81]:
df['dayTripDuration'].dropna(inplace=True)
df['nightTripDuration'].dropna(inplace=True)

In [87]:
import scipy.stats
np.random.seed(42)

## KS Test

#### Null Hypothesis:
The samples of trip durations for Night and Day trips are drawn from the same continuous distribution


In [90]:
scipy.stats.ks_2samp(df.dayTripDuration, df.nightTripDuration)

Ks_2sampResult(statistic=0.026644370156998387, pvalue=2.3133735253554461e-08)

The p-value from this test is 2.313e-08, which is far smaller than my alpha of 0.05, so we reject the null hypothesis that the two samples are drawn from the same continuous distributions

## KS Test (small samples)

#### Null Hypothesis:
The samples of trip durations for Night and Day trips are drawn from the same continuous distribution

In [93]:
dayTripDuration_Small = df['dayTripDuration'].sample(random_state=42, frac=0.005)
nightTripDuration_Small = df['nightTripDuration'].sample(random_state=42, frac=0.005)

In [94]:
scipy.stats.ks_2samp(dayTripDuration_Small, nightTripDuration_Small)

Ks_2sampResult(statistic=0.10968379446640314, pvalue=0.40015718279626111)

The p-value from this test is 0.400, which is larger than my chosen alpha, so we fail to reject the null hypothesis that the two sample are drawn from the same continuous distributions.

## Pearson
### Null Hypothesis:
There is no correlation between Day Trip Duration and Night Trip Duration

In [110]:
# Samples must be broadcast together, sorted, and normalized

day = df.dayTripDuration.dropna()
night = df.nightTripDuration.dropna()

print(len(day))
print(len(night))

36862
19712


In [111]:
day_samp = np.random.choice(day, size=19000)
night_samp = np.random.choice(night, size=19000)

day_samp = np.sort(day_samp)
night_samp = np.sort(night_samp)

In [112]:
def normalize(num, mean, std):
    """normalizes a number"""
    return (num - mean)/std

In [113]:
normalize_vect = np.vectorize(normalize)

day_samp_mean = np.mean(day_samp)
day_samp_std = np.std(day_samp)

night_samp_mean = np.mean(night_samp)
night_samp_std = np.std(night_samp)

day_samp_norm = normalize_vect(day_samp, day_samp_mean, day_samp_std)
night_samp_norm = normalize_vect(night_samp, night_samp_mean, night_samp_std)

In [114]:
scipy.stats.pearsonr(day_samp_norm, night_samp_norm)

(0.99954074286313732, 0.0)

The p-value is 0.0, which is smaller than the alpha of 0.05, so we reject the null hypothesis that there is no correlation between Day Trip Duration and Night Trip Duration

## Spearman
### Null Hypothesis: 
There is no correlation between Day Trip Duration and Night Trip Duration

In [116]:
scipy.stats.spearmanr(day_samp, night_samp)

SpearmanrResult(correlation=0.99999680313673589, pvalue=0.0)

The p-value is 0.0, which is smaller than the alpha of 0.05, so we reject the null hypothesis that there is no correlation between Day Trip Duration and Night Trip Duration